src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java

```diff
@@ -288,18 +290,18 @@ private static AggregatingArtifactValue createAggregatingValue(
         return null;
       }
       if (inputValue instanceof FileArtifactValue) {
-        inputs.add(Pair.of(input, (FileArtifactValue) inputValue));
+        fileInputs.add(Pair.of(input, (FileArtifactValue) inputValue));
       } else if (inputValue instanceof TreeArtifactValue) {
-        inputs.add(Pair.of(input, ((TreeArtifactValue) inputValue).getSelfData()));
+        directoryInputs.add(Pair.of(input, (TreeArtifactValue) inputValue));
       } else {
         // We do not recurse in aggregating middleman artifacts.
         Preconditions.checkState(!(inputValue instanceof AggregatingArtifactValue),
             "%s %s %s", artifact, action, inputValue);
       }
     }
     return (action.getActionType() == MiddlemanType.AGGREGATING_MIDDLEMAN)
```

Comment:
There is probably a Very Good Reason why runfiles middlemen are treated differently from other middlemen, but I can't tell right now and even if that's not necesssary, this change is definitely not the place to fix that. `ActionExecutionFunction.java:694` tells *what* we do with it (we don't put them into `expandedArtifacts`) but it doesn't tell *why* :(

Reply:
The reasons are explained in the commit message of c868c47eaa1.